Docs: Explain Git Sync resource and repository limits (#127829)
grafana/grafana: docs/sources/as-code/observability-as-code/git-sync/usage-limits.md

| Amount of repositories                    | 1                | 10                | 10              | 10                     |
| Amount of synced resources per repository | 20               | 1,000             | No limit        | No limit               |

> **We don't recommend syncing more than 1,000 resources per repository connection today.** This isn't an arbitrary cap: beyond roughly 1,000 resources per connection, the sync workflow puts noticeable load on Grafana itself, and you may see slower syncs and increased database load. We're actively improving this and expect to raise the recommended ceiling in upcoming releases.

Reviewer: ```suggestion
**Do not sync more than 1,000 resources per repository connection as of today.** This isn't an arbitrary cap: beyond roughly 1,000 resources per connection, the sync workflow puts noticeable load on Grafana itself, which may result in slower syncs and increased database load. In any case, Git Sync is under continuos development and the recommended ceiling will increase in upcoming releases.
```

The limit of 10 repository connections is a per-stack limit.

For details on usage and storage limits, refer to [Dashboard and folder limits](https://grafana.com/docs/grafana-cloud/cost-management-and-billing/manage-invoices/understand-your-invoice/usage-limits/#other-usage-limits).

### Scale beyond 1,000 resources per repository

If a single repository holds more than 1,000 resources, you don't have to raise the per-repository limit. Instead, connect the same repository multiple times, with each connection pointing to a different folder (path) in the repository. Each connection syncs its own subset of resources and counts toward the 1,000-resource recommendation independently.

> Because a stack allows up to 10 repository connections, this lets you sync up to roughly 10,000 resources from a single monorepo while keeping each connection within the recommended range. If you need a bit more than that, the 10-connection limit can be increased slightly for Cloud stacks on request — contact Support to discuss your use case.

Reviewer: ```suggestion
Because a stack allows up to 10 repository connections, this lets you sync up to roughly 10,000 resources from a single monorepo while keeping each connection within the recommended range. 

If you need more than that, the 10-connection limit can be increased slightly for Cloud stacks on request. Contact Support to discuss your use case.
```

On Grafana Cloud, syncing this many resources may also exceed your stack's maximum number of dashboards. If it does, you'll need to increase that limit as well. Refer to [Dashboard and folder limits](https://grafana.com/docs/grafana-cloud/cost-management-and-billing/manage-invoices/understand-your-invoice/usage-limits/#other-usage-limits) to review and adjust your dashboard limits.

> When you split a monorepo across several connections, use [folderless sync](https://grafana.com/docs/grafana/<GRAFANA_VERSION>/as-code/observability-as-code/git-sync/key-concepts#sync-targets) for each connection. With folder sync (the default), every connection creates its own wrapper folder named after the repository, so splitting the monorepo would change your folder hierarchy and nest resources one level deeper. Folderless sync places each connection's resources at the top level, so the split has no effect on how your dashboards and folders are organized in Grafana.

Reviewer: I'd move this under line 61, before "If you need more than that...". Keep all the explanation about splitting connections together.

### Modify your usage limits

Before changing your usage limits, study your specific use case. Design the repository structure carefully, and determine how many repositories and how many resources you can support. For example, setting over 1,000 resources per repository may impact your system's performance.

> Limit increases aren't granted automatically. When you request one, we'll ask you to describe your use case so we can understand why the current limits aren't enough, and assess whether the increase is necessary and safe for your stack's performance. In many cases, splitting a monorepo across multiple connections (see [Scale beyond 1,000 resources per repository](#scale-beyond-1000-resources-per-repository)) is a better option than raising the limits.

Reviewer 2: This applies to Cloud, am I right? in On-prem the customer can simply increase the value in the .ini.

Let's make it obvious 👍🏼

Author: Good catch — clarified: the request-and-review flow now explicitly applies to Grafana Cloud, and on-prem users change the limits directly via `max_repositories` / `max_resources_per_repository`.

> Limit increases aren't granted automatically. When you request one, we'll ask you to describe your use case so we can understand why the current limits aren't enough, and assess whether the increase is necessary and safe for your stack's performance. In many cases, splitting a monorepo across multiple connections (see [Scale beyond 1,000 resources per repository](#scale-beyond-1000-resources-per-repository)) is a better option than raising the limits.

Reviewer: ```suggestion
Limit increases aren't granted automatically. When you request one, Support will ask you to describe your use case to understand why the current limits aren't enough, and assess whether the increase is necessary and safe for your stack's performance. In many cases, splitting a monorepo across multiple connections, as described in [Scale beyond 1,000 resources per repository](#scale-beyond-1000-resources-per-repository), is a better option than raising the limits.
```

| Amount of repositories                    | 1                | 10                | 10              | 10                     |
| Amount of synced resources per repository | 20               | 1,000             | No limit        | No limit               |

**Do not sync more than 1,000 resources per repository connection as of today.** This isn't an arbitrary cap: beyond roughly 1,000 resources per connection, the sync workflow puts noticeable load on Grafana itself, which may result in slower syncs and increased database load. In any case, Git Sync is under continuous development and the recommended ceiling will increase in upcoming releases.

The limit of 10 repository connections is a per-stack limit.

> Both of these limits are early figures. We plan to raise them by orders of magnitude as Git Sync matures: we're working toward supporting around 100 repository connections per stack, and up to roughly 1,000 in the longer term, with similar increases to the number of resources you can sync per connection.

Reviewer: We should avoid the use of "we"  as per style guide - https://grafana.com/docs/writers-toolkit/write/style-guide/style-conventions/#address-users-clearly

> Both of these limits are early figures. We plan to raise them by orders of magnitude as Git Sync matures: we're working toward supporting around 100 repository connections per stack, and up to roughly 1,000 in the longer term, with similar increases to the number of resources you can sync per connection.

Reviewer: ```suggestion
Both of these limits are early figures. As Git Sync matures, these limits will raise by orders of magnitude. The goal is for Git Sync to support around 100 repository connections per stack, and up to roughly 1,000 in the longer term, with similar increases to the number of resources you can sync per connection.
```

> If these limits are affecting your use of Git Sync, let us know through the [Grafana roadmap portal](https://grafana.ideas.aha.io/ideas). Open or vote on a feature request for more repository connections or more resources per repository to prioritize it.

Reviewer: ```suggestion
If these limits are affecting your use of Git Sync, [get in touch](https://grafana.com/docs/grafana/<GRAFANA_VERSION>/developer-resources/contribute/#communicate-with-grafana) and explain your situation, or share any idea or suggestion. 
```

> If these limits are affecting your use of Git Sync, let us know through the [Grafana roadmap portal](https://grafana.ideas.aha.io/ideas). Open or vote on a feature request for more repository connections or more resources per repository, so we can prioritize the improvements that matter most to you.

Reviewer: ```suggestion
If these limits are affecting your use of Git Sync, let us know through the [Grafana roadmap portal](https://grafana.ideas.aha.io/ideas). Open or vote on a feature request for more repository connections or more resources per repository to prioritize it.
```

Instead, shard by capacity. Create one repository and group teams into a small number of shard folders, each holding up to about 1,000 resources, and connect each shard folder separately. For example, a customer with 190 teams and 900 resources fits comfortably in a single shard today:

```
your-org/grafana-manifests/
├── shard-1/        ← ~900 resources today, connected now
│   ├── team-a/
│   ├── team-b/
│   └── ...
├── shard-2/        ← empty for now, add a connection when shard-1 approaches 1,000
├── shard-3/        ← add later as you keep growing
└── shard-4/        ← ...up to 10 shards / connections per stack
```

As the number of resources grows, add `shard-2`, `shard-3`, and later shards, and connect each one. You can move teams between shards at any time to balance the load, so you only pay for the connections you actually need and can grow up to the 10-connection limit without restructuring your repository.

> If sharding isn't practical for your setup, we can also raise the per-connection resource limit modestly — for example from 1,000 to around 1,200 or 1,500. This is a small adjustment for a bit of extra headroom, not an order-of-magnitude increase: we don't recommend going well beyond 1,500 resources per connection because of the performance impact on Grafana. For substantially larger scale, sharding remains the recommended approach.

Reviewer: ```suggestion
If sharding isn't practical for your setup, try raising the per-connection resource limit modestly, for example from 1,000 to a limit in the 1,200 - 1,500 range. This is a small adjustment for a bit of extra headroom, not an order-of-magnitude increase. Still, do not go beyond 1,500 resources per connection because of the performance impact on Grafana. For substantially larger scales, sharding remains the recommended approach.
```
